src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5229,48 +5230,42 @@ public static Task<TResult> FromResult<TResult>(TResult result)
                 // null reference types and default(Nullable<T>)
                 return Task<TResult>.s_defaultResultTask;
             }
-            else if (typeof(TResult).IsValueType) // help the JIT avoid the value type branches for ref types
+
+            // For Boolean, we cache all possible values.
+            if (typeof(TResult) == typeof(bool)) // only the relevant branches are kept for each value-type generic instantiation
             {
-                // For Boolean, we cache all possible values.
-                if (typeof(TResult) == typeof(bool)) // only the relevant branches are kept for each value-type generic instantiation
-                {
-                    bool value = (bool)(object)result!;
-                    Task<bool> task = value ? TaskCache.s_trueTask : TaskCache.s_falseTask;
-                    return Unsafe.As<Task<TResult>>(task); // UnsafeCast avoids type check we know will succeed
-                }
-                // For Int32, we cache a range of common values, [-1,9).
-                else if (typeof(TResult) == typeof(int))
+                Task<bool> task = *(bool*)&result ? TaskCache.s_trueTask : TaskCache.s_falseTask;
```

Comment:
Because `(bool)(object)result` looks like it boxes even though we know the JIT optimizes away that boxing. :)

Reply:
I see. I assume in this case it doesn't matter but in general jit hates "&" because it marks argument as "address taken" and then can give up on potential optimizations 🙂

Reply:
You think I should revert it?

Reply:
>  We take the address later on for the same local.

Then it does not matter. Address-taken-ness of the kind Egor mentioned is a per-local "global".

Reply:
Yep, that's exactly why I said "I assume in this case it doesn't matter" 🙂 just wanted to note it in general